bootloader/firmware/updater.c

```diff
@@ -0,0 +1,99 @@
+#include "platform_specific_hal_conf.h"
+#include "bootloader/core/updater.h"
+#include "bootloader/core/util.h"
+#include "bootloader/firmware/constants.h"
+
+
+/**
+ * Callback to buffer iterator that writes a value to flash.
+ * @param address The address to write to
+ * @param data The 64bit value
+ * @return true on success
+ */
+static bool fw_write_to_flash(uint32_t address, uint64_t data);
+
+
+typedef struct {
+    /** Number of data messages received. */
+    uint32_t num_messages_received;
+    /** Running error detection value of update date. */
+    uint32_t error_detection;
+    bool erased;
+} UpdateState;
+
+static UpdateState update_state = {
+    .num_messages_received=0,
+    .error_detection=0,
+    .erased=false
+};
```

Comment:
can we make this "c object-oriented" with a `UpdateState* get_state()` and have the other functions take an `UpdateState*`?